src/System.Text.Utf8String/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,782 @@
+                {
+                    return i;
+                }
+            }
+
+            return StringNotFound;
+        }
+
+        // TODO: Should this be public?
+        public int IndexOf(uint codePoint)
+        {
+            CodePointEnumerator it = GetCodePointEnumerator();
+            while (it.MoveNext())
+            {
+                if (it.Current == codePoint)
+                {
+                    return it.PositionInCodeUnits;
+                }
+            }
+
+            return StringNotFound;
+        }
+
+        // TODO: Naive algorithm, reimplement faster - implemented to keep parity with IndexOf
+        public int LastIndexOf(Utf8Span value)
+        {
+            // Special case for looking for empty strings
+            if (value.Length == 0)
+            {
+                // Maintain parity with .NET C#'s LastIndexOf
+                return Length == 0 ? 0 : Length - 1;
+            }
+
+            if (Length == 0)
+            {
+                return StringNotFound;
+            }
+
+            Utf8Span restOfTheString = this;
+
+            for (int i = Length - 1; i >= value.Length - 1; restOfTheString = Substring(0, i--))
+            {
+                int pos = restOfTheString.LastIndexOf(value[value.Length - 1]);
+                if (pos == StringNotFound)
+                {
+                    return StringNotFound;
+                }
+
+                int substringStart = pos - (value.Length - 1);
+                if (IsSubstringAt(substringStart, value))
+                {
+                    return substringStart;
+                }
+
+            }
+
+            return StringNotFound;
+
+        }
+
+        public int LastIndexOf(byte codeUnit)
+        {
+            for (int i = Length - 1; i >= 0; i--)
+            {
+                if (codeUnit == this[i])
+                {
+                    return i;
+                }
+            }
+
+            return StringNotFound;
+        }
+
+        public int LastIndexOf(uint codePoint)
+        {
+            CodePointReverseEnumerator it = CodePoints.GetReverseEnumerator();
+            while (it.MoveNext())
+            {
+                if (it.Current == codePoint)
+                {
+                    // Move to beginning of code point
+                    it.MoveNext();
+                    return it.PositionInCodeUnits;
+                }
+            }
+
+            return StringNotFound;
+        }
+
+        // TODO: Re-evaluate all Substring family methods and check their parameters name
+        public bool TrySubstringFrom(Utf8Span value, out Utf8Span result)
+        {
+            int idx = IndexOf(value);
+
+            if (idx == StringNotFound)
+            {
+                result = default;
+                return false;
+            }
+
+            result = Substring(idx);
+            return true;
+        }
+
+        public bool TrySubstringFrom(byte codeUnit, out Utf8Span result)
```

Comment:
Yeah, the code unit APIs like that would be very buggy, i.e. a random code unit could match.

Reply:
I think it's ok to leave it for now and do a holistic cleanup of the index based APIs.